### DISCUSSION TITLE
Change content of uiCombo or......

### DISCUSSION THREAD

**Discussion by DRBAUS** [RomanKubiak/ctrlr#137] [General]
Hi all,

I am going to ask again. Spending most of the day and night figuring out how to do this but without any working results. 

Is it possible to change the content of a uiCombo? I have a wavelist setup now with waveform 1-72 but it would be nice to be able to input a user waveform name. 

Another option would be to have a value slider, an LCD Display and 72 labels. The LCD will then display one of the labels contents which is selected by the value slider. 

I have been able to show Combo values on a display but that's a different method.

As usual I am clueless and thankful for the pointers in the right direction.

BAUS

**Comment by DRBAUS** [comment]
I always like it when asking a question and then the answer pops up right after. 

This will do the trick for option 2

`WAVEFORMSELECT = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

label1 = L(panel:getComponent("JD-990"):getText())
label2 = L(panel:getComponent("K2000"):getText()) 

	
	if value==1 then

		panel:getComponent("WAVEFORMDISPLAY"):setProperty("uiLabelText", label1, false)


	elseif value==2 then

		panel:getComponent("WAVEFORMDISPLAY"):setProperty("uiLabelText", label2, false)

end
end
`

**Comment by Ted-juh** [comment]
I use a code that deals with dynamically updating the content of a combobox from a StringArray. The content of a StringArray can be altered like a table in lua. So, to get a working method for option 1, I'm rewriting the code to deal with user input from a label. I can insert a string at a given position in the StringArray but that also means that the data of the original position gets shifted one place down. Still working out how to deal with that.

Edit: Nvm, I can use remove/ removeString. 

Code follows later.

**Comment by dobo365** [comment]
Hi,
  I'm doing that in some of my panels especially on the Prologue one where the default user osc names can be replaced by user ones.

Different things:
- at panel load you should initialize all your modulators in a method like AssignModulators
```
	-- Program
	modProgramLevel = panel:getModulatorByName("ProgramLevel")
	modProgramTranspose = panel:getModulatorByName("ProgramTranspose")
	modProgramTuning = panel:getModulatorByName("ProgramTuning")
	modAmpVelocity = panel:getModulatorByName("AmpVelocity")
	modModEffectType = panel:getModulatorByName("ModEffectType")
	modModEffectSubType = panel:getModulatorByName("ModEffectSubType")
```
This avoids Ctrlr to search for each modulator when using panel:getModulatorByName() everywhere
In your code you then just do 
`		modDelayReverb:setValue(1, true)`
or 
`		modDelayReverbType:getComponent():setPropertyString ("uiComboContent", sTemp)`

Then to your issue (you have the formula above)
- put the original list of your combo in a string like
` 		sTemp = "STEREO=0\nMONO=7\nPING PONG=13\nHIPASS=20\nTAPE=26\nONE TAP=32\nSTEREO BPM=39\nMONO BPM=45\nPING BPM=52\nHIPASS BPM=58\nTAPE BPM=64\nDOUBLING=71\n"`
- the =7 for example is when you need specific a midi value for that combo value
- change the string at any time (in my case I can switch the ModEffectSubType based on the selection of ModEffectType) by using modDelayReverbType:getComponent():setPropertyString ("uiComboContent", sTemp)

And another thing:
if you use 
`modXXX:getComponent():setProperty("uiLabelText", label2, false)`
or 	
`modXXX:getComponent():setPropertyString ("uiLabelText", "")`
you should use its equivalent
`modXXX:getComponent():getProperty("uiLabelText")`
So, no need to use 
`label1 = L(panel:getComponent("JD-990"):getText())`

But of course, this is programming and each of us decides its own way of doing things ;-)

**Comment by Ted-juh** [comment]
Not without its flaws but here is a panel that should help with option 1. 
[Combotest.zip](https://github.com/RomanKubiak/ctrlr/files/5827048/Combotest.zip)

One thing I don't get:

I do setSelectedId in the init file, there is no error but it doesn't set it to that position of the combobox. So it results in an error when opening the panel because in the test method it asks for the getSelectedItemIndex which is now nil (or -1 actually). What am I doing wrong? And Baus, the array gets overwritten every time you open the panel, a workaround could be to save it to and read it from a file.
